flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncWaitOperator.java

```diff
@@ -293,140 +249,113 @@ public void close() throws Exception {
 			waitInFlightInputsFinished();
 		}
 		finally {
-			Exception exception = null;
-
-			try {
-				super.close();
-			} catch (InterruptedException interrupted) {
-				exception = interrupted;
-
-				Thread.currentThread().interrupt();
-			} catch (Exception e) {
-				exception = e;
-			}
-
-			try {
-				// terminate the emitter, the emitter thread and the executor
-				stopResources(true);
-			} catch (InterruptedException interrupted) {
-				exception = ExceptionUtils.firstOrSuppressed(interrupted, exception);
-
-				Thread.currentThread().interrupt();
-			} catch (Exception e) {
-				exception = ExceptionUtils.firstOrSuppressed(e, exception);
-			}
-
-			if (exception != null) {
-				LOG.warn("Errors occurred while closing the AsyncWaitOperator.", exception);
-			}
+			super.close();
 		}
 	}
 
-	@Override
-	public void dispose() throws Exception {
-		Exception exception = null;
+	/**
+	 * Add the given stream element to the operator's stream element queue. This operation blocks until the element
+	 * has been added.
+	 *
+	 * <p>Between two insertion attempts, this method yields the execution to the mailbox, such that events as well
+	 * as asynchronous results can be processed.
+	 *
+	 * @param streamElement to add to the operator's queue
+	 * @throws InterruptedException if the current thread has been interrupted while yielding to mailbox
+	 * @return a handle that allows to set the result of the async computation for the given element.
+	 */
+	private ResultFuture<OUT> addToWorkQueue(StreamElement streamElement) throws InterruptedException {
+		assert(Thread.holdsLock(checkpointingLock));
 
-		try {
-			super.dispose();
-		} catch (InterruptedException interrupted) {
-			exception = interrupted;
+		pendingStreamElement = streamElement;
 
-			Thread.currentThread().interrupt();
-		} catch (Exception e) {
-			exception = e;
+		Optional<ResultFuture<OUT>> queueEntry;
+		while (!(queueEntry = queue.tryPut(streamElement)).isPresent()) {
+			mailboxExecutor.yield();
```

Comment:
How to guarantee that `yield()` could consume the elements from `queue` finally?

In previous way the `EmitterThread` could consume the queue, but now I did not see any runnable to be submitted into mailbox for consuming the queue. I know `outputCompletedElements()` could empty the queue, but it can not happen before calling `userFunction.asyncInvoke(element.getValue(), resultHandler)`?

Reply:
`outputCompletedElements` is called after an element has been completed by the asyncUserFunction. Only at that point, the queue entries have actually a changed state. There is no other code path that can actually make any queue entry completed. So consuming the queue at any other location will not yield a different result.

The only other way an item can be actually be completed is if we insert a watermark into an empty queue, as watermarks are always completed.